### PR TITLE
Disable autodownload by default

### PR DESCRIPTION
**Pull request by shrit**
Signed-off-by: Omar Shrit <omar@shrit.me>

### CODE REVIEW

**Review by rcurtin — COMMENTED** [review]
I agree with the idea here, but I think we'll need to make a few CI changes.  ensmallen will need to be installed; I think you should be able to use `apt-get`, `brew`, and `nuget` in the various `.ci/` files.  There may be a bit of work to be done on the build nodes on Jenkins, but that should be pretty easy.

**Comment by shrit** [comment]
I will update the `ci` to download all necessary dependencies :+1:

**Comment by rcurtin** [comment]
I installed `libensmallen-dev` and `libstb-dev` on all the CI build nodes I control, so hopefully the Jenkins jobs should now work correctly too.  @zoq would you mind installing those packages on the CI nodes of yours? :)

**Comment by zoq** [comment]
Some or most of the jobs use https://github.com/mlpack/jenkins-conf/blob/master/Dockerfiles/jenkins-amd64-debian.Dockerfile so we only have to update the file and rebuild the docker. I'll update it in a second and push it.

**Comment by zoq** [comment]
Docker files updated, and images rebuild.

File: .ci/linux-steps.yaml [hunk 1]
```diff
@@ -44,7 +44,12 @@ steps:
 
     # Install cereal.
     wget https://github.com/USCiLab/cereal/archive/v1.3.0.tar.gz
-    tar -xvzpf v1.3.0.tar.gz # Unpack into cereal-1.3.0/.
+        tar -xvzpf v1.3.0.tar.gz # Unpack into cereal-1.3.0/.
+        cd cereal-1.3.0 && \
+        cmake . && \
+        make && \
```
Comment (rcurtin) [inline .ci/linux-steps.yaml:50]: You can probably get away without the `make` here---since cereal is header-only, I guess we don't need to build anything.  In fact, you could even do something just like `cp -vr include/cereal /usr/include/` or something like that.  I think that will help with the Linux build errors, at least. :)

**Comment by shrit** [comment]
I found it strange that the CI was not able to locate `libensmallen-dev` maybe I did not write the package name correctly? it seems to be a part of the Ubuntu repository.
Also, there is a lib `libstb-dev` available for Ubuntu and Debian. Let me know if it has the same headers or, I will just wget it from mlpack website.

**Comment by rcurtin** [comment]
It looks like the Azure build workers use xenial, which is pretty old and doesn't have `libensmallen-dev` available (you can see this in the output, but the build doesn't fail). It seems like the easiest solution would be to just install ensmallen from source too. :+1:

File: .ci/linux-steps.yaml [hunk 2]
```diff
@@ -42,9 +42,30 @@ steps:
...
+        sudo cp -vr include/* /usr/include/ && \
+        cd ..
+
     # Install cereal.
     wget https://github.com/USCiLab/cereal/archive/v1.3.0.tar.gz
-    tar -xvzpf v1.3.0.tar.gz # Unpack into cereal-1.3.0/.
+        tar -xvzpf v1.3.0.tar.gz # Unpack into cereal-1.3.0/.
+        cd cereal-1.3.0 && \
+        cmake . && \
+        make && \
```
Comment (rcurtin) [inline .ci/linux-steps.yaml:66]:
```suggestion
```
I think you can just remove these; we can install directly without building.

File: .ci/linux-steps.yaml [hunk 3]
```diff
@@ -42,9 +42,30 @@ steps:
         sudo make install && \
         cd ..
 
+    # Install ensmallen.
+    wget https://ensmallen.org/files/ensmallen-latest.tar.gz
+        tar -xvzpf ensmallen-latest.tar.gz # Unpack into ensmallen-*/.
+        cd ensmallen-* && \
+        cmake . && \
+        make && \
```
Comment (rcurtin) [inline .ci/linux-steps.yaml:50]:
```suggestion
```
ensmallen is header-only too, so unless I am forgetting something, I think you can just install directly too. :+1:

**Review by rcurtin — COMMENTED** [review]
Just left two comments that should hopefully fix the build. :+1:

File: .ci/macos-steps.yaml [hunk 4]
```diff
@@ -14,7 +14,7 @@ steps:
     set -e
     sudo xcode-select --switch /Applications/Xcode_12.2.app/Contents/Developer
     unset BOOST_ROOT
-    brew install libomp openblas armadillo boost cereal
+    brew install libomp openblas armadillo boost cereal ensmallen
```
Comment (rcurtin) [inline .ci/macos-steps.yaml:17]: I guess stb will need to be installed too?

File: .ci/windows-steps.yaml [hunk 5]
```diff
@@ -14,6 +14,7 @@ steps:
     nuget install boost_math_c99-vc140 -o $(Agent.ToolsDirectory) -Version 1.60.0
     nuget install OpenBLAS -o $(Agent.ToolsDirectory)
     nuget install unofficial-flayan-cereal -o $(Agent.ToolsDirectory)
+    vcpkg install ensmallen:x64-windows
```
Comment (rcurtin) [inline .ci/windows-steps.yaml:17]: I'm not sure if this installed to the right place---at least, CMake wasn't able to find it.  I wonder if you can use nuget to install instead?  I'm not sure, I don't know much about Windows configuration... I assume also STB will be necessary here too.

Comment (shrit) [reply .ci/windows-steps.yaml:17]: The same as you, I have no clue on what is happening on the Windows side, I will use `nuget` and see what happens, for stb, I think we need to download it too.

File: .ci/macos-steps.yaml [hunk 6]
```diff
@@ -25,6 +25,13 @@ steps:
       brew install --cask julia
     fi
 
+    # Install STB.
+    wget https://mlpack.org/files/stb.tar.gz
+        tar -xvzpf stb.tar.gz # Unpack into stb/.
+        cd stb &&\
+        sudo cp -vr include/* /usr/include/ && \
```
Comment (rcurtin) [inline .ci/macos-steps.yaml:32]: I wonder if maybe `/usr/include/` isn't the right path to install to on OS X?

File: .ci/windows-steps.yaml [hunk 7]
```diff
@@ -12,8 +12,8 @@ steps:
     nuget install boost -o $(Agent.ToolsDirectory) -Version 1.60.0
     nuget install boost_random-vc140 -o $(Agent.ToolsDirectory) -Version 1.60.0
     nuget install boost_math_c99-vc140 -o $(Agent.ToolsDirectory) -Version 1.60.0
-    nuget install OpenBLAS -o $(Agent.ToolsDirectory)
     nuget install unofficial-flayan-cereal -o $(Agent.ToolsDirectory)
+    nuget install ensmallen:x64-windows -o $(Agent.ToolsDirectory)
```
Comment (rcurtin) [inline .ci/windows-steps.yaml:16]: Maybe the `:x64-windows` isn't needed here?  I am not sure.

File: .ci/windows-steps.yaml [hunk 8]
```diff
@@ -12,8 +12,8 @@ steps:
     nuget install boost -o $(Agent.ToolsDirectory) -Version 1.60.0
     nuget install boost_random-vc140 -o $(Agent.ToolsDirectory) -Version 1.60.0
     nuget install boost_math_c99-vc140 -o $(Agent.ToolsDirectory) -Version 1.60.0
-    nuget install OpenBLAS -o $(Agent.ToolsDirectory)
```
Comment (rcurtin) [inline .ci/windows-steps.yaml:15]: Any particular reason to remove this?  Seems like we should keep it installed.

File: .ci/windows-steps.yaml [hunk 9]
```diff
@@ -62,6 +62,7 @@ steps:
     -DBOOST_INCLUDEDIR=$(Agent.ToolsDirectory)\boost.1.60.0.0\lib\native\include `
     -DBOOST_LIBRARYDIR=$(Agent.ToolsDirectory)\boost_libs `
     -DCEREAL_INCLUDE_DIR=$(Agent.ToolsDirectory)\unofficial-flayan-cereal.1.2.2\build\native\include `
+    -DENSMALLEN_INCLUDE_DIR=$(Agent.ToolsDirectory)\ensmallen.2.16.2\include`
```
Comment (rcurtin) [inline .ci/windows-steps.yaml:65]: I think this will go out of date as the ensmallen package changes version, unless we hardcode the version earlier or detect the version here or something.

File: .ci/windows-steps.yaml (hunk 8 above)
Comment (shrit) [reply .ci/windows-steps.yaml:15]: It is a repeated line, it was first mentioned in line 10 or 11.

File: .ci/macos-steps.yaml (hunk 6 above)
Comment (shrit) [reply .ci/macos-steps.yaml:32]: @zoq Any thoughts? I have no idea how OS X is structured internally

File: .ci/windows-steps.yaml (hunk 8 above)
Comment (rcurtin) [reply .ci/windows-steps.yaml:15]: Ahh, ok; I did not expand the lines to see that. :smile:

File: .ci/windows-steps.yaml (hunk 9 above)
Comment (shrit) [reply .ci/windows-steps.yaml:65]: Agreed, we need to find a way to encode the version. In the meanwhile, let us see if we were able to install ensmallen at least.

File: .ci/windows-steps.yaml (hunk 7 above)
Comment (rcurtin) [reply .ci/windows-steps.yaml:16]: Let me know if you want me to try and debug this one.  I am sure the build will work right; we just have to convince Windows to download ensmallen correctly, and then MSVC to look in the correct path. :)

Comment (shrit) [reply .ci/windows-steps.yaml:16]: Agreed, ensmallen is not in `nuget` I have opened an issue in ensmallen, [here](https://github.com/mlpack/ensmallen/issues/311), but no one volunteered, I have been looking into it last night, and it seems to me (not sure yet) that it might be possible to use Linux to create a nuget package.
I will give a try tomorrow if I have a spare time.

Comment (rcurtin) [reply .ci/windows-steps.yaml:16]: It is available in vcpkg though (https://github.com/microsoft/vcpkg/issues/5015), so maybe we can manage to use that instead?  I can't remember if we already tried it or what the outcome there was.

File: .ci/macos-steps.yaml (hunk 6 above)
Comment (rcurtin) [reply .ci/macos-steps.yaml:32]: The build logs indicate that other libraries are found in `/usr/local/include/`, so maybe you should try there?

```
2021-08-07T17:23:36.2370180Z -- Found cereal: /usr/local/include (found suitable version "1.3.0", minimum required is "1.1.2") 
2021-08-07T17:23:36.7662720Z -- Found Boost: /usr/local/include (found suitable version "1.76.0", minimum required is "1.58")  
```

File: .ci/windows-steps.yaml [hunk 10]
```diff
@@ -62,6 +62,7 @@ steps:
     -DBOOST_INCLUDEDIR=$(Agent.ToolsDirectory)\boost.1.60.0.0\lib\native\include `
     -DBOOST_LIBRARYDIR=$(Agent.ToolsDirectory)\boost_libs `
     -DCEREAL_INCLUDE_DIR=$(Agent.ToolsDirectory)\unofficial-flayan-cereal.1.2.2\build\native\include `
+    -DENSMALLEN_INCLUDE_DIR=$(Agent.ToolsDirectory)\ensmallen.2.17.0\include`
```
Comment (rcurtin) [inline .ci/windows-steps.yaml:65]:
```suggestion
    -DENSMALLEN_INCLUDE_DIR=$(Agent.ToolsDirectory)\ensmallen.2.17.0\include `
```
I think that there has to be a space before the line wrap character... just guessing. :)

File: .ci/windows-steps.yaml [hunk 11]
```diff
@@ -12,8 +12,8 @@ steps:
     nuget install boost -o $(Agent.ToolsDirectory) -Version 1.60.0
     nuget install boost_random-vc140 -o $(Agent.ToolsDirectory) -Version 1.60.0
     nuget install boost_math_c99-vc140 -o $(Agent.ToolsDirectory) -Version 1.60.0
-    nuget install OpenBLAS -o $(Agent.ToolsDirectory)
     nuget install unofficial-flayan-cereal -o $(Agent.ToolsDirectory)
+    nuget install ensmallen -o $(Agent.ToolsDirectory) -Version 2.17.0
```
Comment (rcurtin) [inline .ci/windows-steps.yaml:16]:
```suggestion
dir $(Agent.ToolsDirectory)
```
Maybe this will show us more?

**Comment by zoq** [comment]
@mlpack-jenkins test this please

File: CMakeLists.txt [hunk 12]
```diff
@@ -14,7 +14,7 @@ option(ARMA_EXTRA_DEBUG "Compile with extra Armadillo debugging symbols." OFF)
 option(TEST_VERBOSE "Run test cases with verbose output." OFF)
 option(BUILD_TESTS "Build tests." ON)
 option(BUILD_CLI_EXECUTABLES "Build command-line executables." ON)
-option(DISABLE_DOWNLOADS "Disable downloads of dependencies during build." OFF)
+option(DISABLE_DOWNLOADS "Disable downloads of dependencies during build." ON)
```
Comment (rcurtin) [inline CMakeLists.txt:17]: To fix the Jenkins build, we'll at least need to make sure that `STB_IMAGE_INCLUDE_DIR` is only appended to `MLPACK_INCLUDE_DIRS` in the block `if (STB_IMAGE_FOUND)`.

Comment (conradsnicta) [inline CMakeLists.txt:17]: The use of a double negative here can be confusing -- using "ON" to actually disable something.  

For ease of understanding (and easier maintenance) it's generally best to have a simple on/off for a given functionality.  I recommend to change the option to `ENABLE_DOWNLOADS` or more accurately `DOWNLOAD_DEPENDENCIES`.  

Using "ON" would then clearly mean to enable downloading the dependencies.

Comment (rcurtin) [reply CMakeLists.txt:17]: Agreed, this would be a nice improvement. :+1:

File: CMakeLists.txt [hunk 13]
```diff
@@ -278,7 +278,7 @@ if (CMAKE_CROSSCOMPILING)
   search_openblas(0.3.13)
 endif()
 
-if (DISABLE_DOWNLOADS)
+if (DOWNLOAD_DEPENDENCIES)
```
Comment (rcurtin) [inline CMakeLists.txt:281]: I'm trying to understand why linking on Windows is failing, but the output doesn't provide much information.  Do you think you can add something like
```
message(STATUS "Armadillo libraries: ${ARMADILLO_LIBRARIES}")
```
down where `MLPACK_LIBRARIES` is set?  That might provide more of a clue of what is going on. :+1:

I dug deeper in the log too to find the linking message:
```
 C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30037\bin\HostX64\x64\link.exe /ERRORREPORT:QUEUE /OUT:"D:\a\1\s\build\Release\mlpack_approx_kfn.exe" /INCREMENTAL:NO /NOLOGO "-static" ..\..\..\..\Release\mlpack.lib C:\hostedtoolcache\windows\OpenBLAS.0.2.14.1\lib\native\lib\x64\libopenblas.dll.a C:\hostedtoolcache\windows\OpenBLAS.0.2.14.1\lib\native\lib\x64\libopenblas.dll.a kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /PDB:"D:/a/1/s/build/Release/mlpack_approx_kfn.pdb" /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:"D:/a/1/s/build/Release/mlpack_approx_kfn.lib" /MACHINE:X64  /machine:x64 mlpack_approx_kfn.dir\Release\approx_kfn_main.obj
 ```
so, it looks like it is linking against OpenBLAS, but compilation is failing because symbols like `_wrapper_ddot` and so forth are not found.  To me, this indicates that compilation was done with `ARMA_USE_WRAPPER` set to true, but then the linking is done as if there is no Armadillo library.

If you print `ARMADILLO_LIBRARIES`, I think that will show us whether or not the Armadillo library has been found.  I suspect that it won't be found---and then maybe we can fix it by modifying the Azure build instructions to specify the `-DARMADILLO_LIBRARY` option when configuring mlpack.  That might work, but that is kind of just a guess...

Comment (shrit) [reply CMakeLists.txt:281]: I will give it a try, I have already spent a couple of hours this week trying to figure out what is happening. I can not see why adding ensmallen result in lost armadillo. The only theory I have is related to armadillo version. It might be magically finding the armadillo headers or library packaged with ensmallen nuget package instead of the version we have already installed.
The only reason made me think of such a theory is related to the configuration when CMake is run, it should print the path of armadillo library, but instead it is  only printing the headers:
```
-- Found Armadillo: D:/a/1/s/armadillo-8.400.0/include (found suitable version "8.400.0", minimum required is "8.400.0")
```
Anyway, I will print `ARMADILLO_LIBRARIES` and see where it will guide us :+1:

Comment (rcurtin) [reply CMakeLists.txt:281]: Yes, I was thinking the same thing.  Another option here might be to forcibly disable the wrapper on Windows by modifying `armadillo_bits/config.hpp` to ensure that `ARMA_USE_WRAPPER` is not set.

Comment (shrit) [reply CMakeLists.txt:281]: It seems that it is not setting the Armadillo library to a specific `.dll` file, the variable has only `openblas files`. Please have a look: https://dev.azure.com/mlpack/mlpack/_build/results?buildId=7193&view=logs&j=934eac7e-378e-5e05-ed87-f05e34fa9fdc&t=9715bafe-ea9f-54fa-e80f-ab012eb34e65&l=45

File: .ci/windows-steps.yaml [hunk 14]
```diff
@@ -62,6 +64,7 @@ steps:
     -DBOOST_INCLUDEDIR=$(Agent.ToolsDirectory)\boost.1.60.0.0\lib\native\include `
```
Comment (rcurtin) [inline .ci/windows-steps.yaml:64]: What happens if you change `\include` to `\tmp\include` for `ARMADILLO_INCLUDE_DIR`?

Comment (rcurtin) [reply .ci/windows-steps.yaml:64]: Or maybe you can just leave `ARMADILLO_LIBRARY` out?

File: .ci/windows-steps.yaml [hunk 15]
```diff
@@ -12,8 +12,10 @@ steps:
...
     nuget install boost_random-vc140 -o $(Agent.ToolsDirectory) -Version 1.60.0
     nuget install boost_math_c99-vc140 -o $(Agent.ToolsDirectory) -Version 1.60.0
-    nuget install OpenBLAS -o $(Agent.ToolsDirectory)
     nuget install unofficial-flayan-cereal -o $(Agent.ToolsDirectory)
+    nuget install ensmallen -o $(Agent.ToolsDirectory) -Version 2.17.0
+    dir $(Agent.ToolsDirectory)
+    tree $(Agent.ToolsDirectory)
 
     mkdir -p $(Agent.ToolsDirectory)/boost_libs
     cp $(Agent.ToolsDirectory)/boost_math_c99-vc140.1.60.0.0/lib/native/address-model-64/lib/*.* $(Agent.ToolsDirectory)/boost_libs
```
Comment (rcurtin) [inline .ci/windows-steps.yaml:21]: I know this was not part of your PR, but I'm pretty sure if you change `tar -xvzf` to `tar -xvf` below, that should download the Armadillo source correctly, and then we can see if using `tmp/include/` as the include directory succeeds. :+1:

Comment (shrit) [reply .ci/windows-steps.yaml:22]: Interesting observation to add to this pull request:
In addition to this pull request, I have added `message(STATUS "ARMADILLO LIBRARY" ${ARMADILLO_LIBRARIES})` in this pull request #3040 
As you can see, it did not print the armadillo library location, but instead it did print the OpenBLAS location

https://dev.azure.com/mlpack/mlpack/_build/results?buildId=7268&view=logs&j=934eac7e-378e-5e05-ed87-f05e34fa9fdc&t=9715bafe-ea9f-54fa-e80f-ab012eb34e65&l=34

I think there is a small bug here in the armadillo deployment on Windows that is passing everywhere, but it detected somehow with this pull request.

Comment (rcurtin) [reply .ci/windows-steps.yaml:22]: Thanks for linking to that.  I looked through the CMake code and I am not entirely sure what is going on, but I think I at least have a partial solution.  Armadillo has two linking modes, controlled by whether the macro `ARMA_USE_WRAPPER` is set or not.  On Windows, if `ARMA_USE_WRAPPER` is set, we must link against Armadillo *and* OpenBLAS; if it is not set, we only need to link against OpenBLAS.

By default, Armadillo ships with a file `armadillo_bits/config.hpp` that leaves `ARMA_USE_WRAPPER` *unset*, in the `include/` directory.  When we run CMake in the Armadillo sources, this generates a directory `tmp/`, and inside that directory there will be a `tmp/include/armadillo_bits/config.hpp` that *will* have `ARMA_USE_WRAPPER` set.

However, I can see in our choice of `ARMADILLO_INCLUDE_DIR` below that we are *not* using the `tmp/` directory, so the mlpack file `CMake/FindArmadillo.cmake` should detect that `ARMA_USE_WRAPPER` is *not* set.  Then, this code runs:

```
if(_ARMA_USE_WRAPPER)
  # UNIX paths are standard, no need to write.
  find_library(ARMADILLO_LIBRARY
    NAMES armadillo
    PATHS "$ENV{ProgramFiles}/Armadillo/lib"  "$ENV{ProgramFiles}/Armadillo/lib64" "$ENV{ProgramFiles}/Armadillo"
    )
    set(_ARMA_REQUIRED_VARS ARMADILLO_LIBRARY ARMADILLO_INCLUDE_DIR VERSION_VAR ARMADILLO_VERSION_STRING)
else()
  # don't link to armadillo in this case
  set(ARMADILLO_LIBRARY "")
endif()
```

So, in our case, that should set `ARMADILLO_LIBRARY` to the empty string, which is why it does not appear in `ARMADILLO_LIBRARIES`.  This code runs at the end of `CMake/FindArmadillo.cmake`:

```
if (ARMADILLO_FOUND)
  set(ARMADILLO_INCLUDE_DIRS ${ARMADILLO_INCLUDE_DIR})
  set(ARMADILLO_LIBRARIES ${ARMADILLO_LIBRARY} ${_ARMA_SUPPORT_LIBRARIES})
endif ()
```

So, we can see based on the CMake output in the log you linked to that we are configuring mlpack as if Armadillo has `ARMA_USE_WRAPPER` *unset*.

But, I am confused, because then when linking we get all kinds of errors like this:

```
2021-09-08T22:50:13.7010984Z   mlpack.lib(lcc.obj) : error LNK2001: unresolved external symbol wrapper2_dgemm_ [D:\a\1\s\build\src\mlpack\tests\mlpack_test.vcxproj]
```

That's confusing because `_wrapper2_dgemm` shouldn't ever appear unless `ARMA_USE_WRAPPER` is set!

So, in conclusion, I'm not sure how `ARMA_USE_WRAPPER` is getting turned on, but I think if you change `ARMADILLO_INCLUDE_DIR` to be `armadillo-9.800.6/tmp/include/` instead of just `armadillo-9.800.6/include/`, then CMake will detect `ARMA_USE_WRAPPER` as set, and then `ARMADILLO_LIBRARIES` should also include `armadillo.lib`.  Maybe do you want to try that and see what it does?

Comment (shrit) [reply .ci/windows-steps.yaml:22]: Done

**Comment by shrit** [comment]
@rcurtin Good news, Armadillo library has been finally found on windows, let us see if there will be no linking issues, if yes then we  can merge this pull request :+1:

**Comment by shrit** [comment]
Here is the last error from this pull request,
@rcurtin, I think this error is coming from the armadillo, any possibilities of am ABI incompatibility?

```
##[error]mlpack.lib(adaboost_model.obj)(0,0): Error LNK2001: unresolved external symbol "class std::mersenne_twister_engine<unsigned __int64,64,312,156,31,-5403634167711393303,29,6148914691236517205,17,8202884508482404352,37,-2270628950310912,43,6364136223846793005> arma::mt19937_64_instance" (?mt19937_64_instance@arma@@3V?$mersenne_twister_engine@_K$0EA@$0BDI@$0JM@$0BP@$0?EKPNJAKFFGJJOGBH@$0BN@$0FFFFFFFFFFFFFFFF@$0BB@$0HBNGHPPPONKGAAAA@$0CF@$0?IBBCAAAAAAAAA@$0CL@$0FIFBPECNEMJFHPCN@@std@@A)
```

**Comment by rcurtin** [comment]
Wait a second... `mt19937_64_instance` is not a symbol that exists in Armadillo 9.800.6!  Is it possible that Armadillo is installed *twice* on the system?  I think this might be the case---is the ensmallen nuget package shipping with a bundled Armadillo?  I think this could cause a problem if *both* `ENSMALLEN_INCLUDE_DIR` and `ARMADILLO_INCLUDE_DIR` have `armadillo_bits/` subdirectories---the compiler might include the wrong one.

If this is correct (please double-check my reasoning above), then I see two options:

1. Remove the Armadillo headers that shipped with ensmallen, so that we are always using Armadillo 9.800.6 in this build.
2. Use the Armadillo headers that shipped with ensmallen, instead of downloading and compiling Armadillo locally.

I think the first solution might be preferable, since it always allows us to build mlpack against the minimum required version of Armadillo.

File: .ci/windows-steps.yaml [hunk 16]
```diff
@@ -12,9 +12,12 @@ steps:
...
     nuget install boost_random-vc140 -o $(Agent.ToolsDirectory) -Version 1.60.0
     nuget install boost_math_c99-vc140 -o $(Agent.ToolsDirectory) -Version 1.60.0
-    nuget install OpenBLAS -o $(Agent.ToolsDirectory)
     nuget install unofficial-flayan-cereal -o $(Agent.ToolsDirectory)
-
+    nuget install ensmallen -o $(Agent.ToolsDirectory) -Version 2.17.0
+    dir $(Agent.ToolsDirectory)
+    tree $(Agent.ToolsDirectory)
+    ## Delete all ensmallen dependencies, we do not need them here
+    Remove-Item $(Agent.ToolsDirectory)\ensmallen.2.17.0\installed\x64-linux\share -Force -Recurse
```
Comment (shrit) [inline .ci/windows-steps.yaml:20]: @rcurtin I have already done that, I suppose that maybe it is not working? I will try to verify if the directory is still there

Comment (rcurtin) [reply .ci/windows-steps.yaml:20]: Ah, sorry, I missed that.  Well, maybe it is worth printing everything in the ensmallen directory, just to make sure the headers are not there.

Comment (shrit) [reply .ci/windows-steps.yaml]: What I find strange is that it has never been able to link with any armadillo version including the one that is in ensmallen package. However, when I corrected the path to add the `tree\include`, it showed behaviour similar as if it was linking with the package.
I have added a tree to see if armadillo is not deleted from the ensmallen package, let us see.

Comment (shrit) [reply .ci/windows-steps.yaml]: @rcurtin we are sure now that visual studio is trying to use the local headers, which is causing all of these issues.
Have a look here
https://dev.azure.com/mlpack/mlpack/_build/results?buildId=7292&view=logs&j=934eac7e-378e-5e05-ed87-f05e34fa9fdc&t=4ee4870b-c6ca-5a83-cdbe-639dd57a3582&l=9973

Comment (rcurtin) [reply .ci/windows-steps.yaml]: Take a look at the error:

```
##[error]C:\hostedtoolcache\windows\ensmallen.2.17.0\installed\x64-linux\include\armadillo(21,10): Error C1083: Cannot open include file: 'armadillo_bits/compiler_check.hpp': No such file or directory
```

I think that you have removed everything of Armadillo that's distributed with ensmallen... *except* the `armadillo` file. So if you remove that too, in addition to the `armadillo_bits/` directory, then I think we will be closer to something that works. :)

**Comment by shrit** [comment]
@rcurtin I think it has passed now, finally, I only got out of heap space error, which is normal from time to time.
I restarted the windows build to see if it may pass this time.

File: .ci/windows-steps.yaml [hunk 17]
```diff
@@ -12,9 +12,14 @@ steps:
...
-    nuget install OpenBLAS -o $(Agent.ToolsDirectory)
     nuget install unofficial-flayan-cereal -o $(Agent.ToolsDirectory)
-
+    nuget install ensmallen -o $(Agent.ToolsDirectory) -Version 2.17.0
+    ## Delete all ensmallen dependencies including armadillo headers, we do not need them here
+    Remove-Item $(Agent.ToolsDirectory)\ensmallen.2.17.0\installed\x64-linux\share -Force -Recurse
+    Remove-Item $(Agent.ToolsDirectory)\ensmallen.2.17.0\installed\x64-linux\include\armadillo_bits -Force -Recurse
+    Remove-Item $(Agent.ToolsDirectory)\ensmallen.2.17.0\installed\x64-linux\include\armadillo -Force
+    dir $(Agent.ToolsDirectory)    
+    tree $(Agent.ToolsDirectory)
```
Comment (rcurtin) [inline .ci/windows-steps.yaml:22]:
```suggestion
```
Now that it works, we can clean this up. :smile:

File: CMakeLists.txt [hunk 18]
```diff
@@ -14,7 +14,7 @@ option(ARMA_EXTRA_DEBUG "Compile with extra Armadillo debugging symbols." OFF)
 option(TEST_VERBOSE "Run test cases with verbose output." OFF)
 option(BUILD_TESTS "Build tests." ON)
 option(BUILD_CLI_EXECUTABLES "Build command-line executables." ON)
-option(DISABLE_DOWNLOADS "Disable downloads of dependencies during build." OFF)
+option(DOWNLOAD_DEPENDENCIES "Download of dependencies during build." OFF)
```
Comment (rcurtin) [inline CMakeLists.txt:17]:
```suggestion
option(DOWNLOAD_DEPENDENCIES "Automatically download dependencies if not available." OFF)
```
Just a minor clarification.

File: CMakeLists.txt [hunk 19]
```diff
@@ -392,6 +392,7 @@ endif()
 set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} ${Boost_INCLUDE_DIRS})
 set(MLPACK_LIBRARIES ${MLPACK_LIBRARIES})
 set(MLPACK_LIBRARY_DIRS ${MLPACK_LIBRARY_DIRS})
+message(STATUS "ARMADILLO LIBRARIES: ${ARMADILLO_LIBRARIES}")
```
Comment (rcurtin) [inline CMakeLists.txt:395]:
```suggestion
```
Just another cleanup now that we know this is working right. :)

**Review by rcurtin — APPROVED** [review]
Everything looks good to me here.  Thank you so much for putting so much effort into this!  Before merge, I think we should still make sure that the Windows build succeeds at least once (we might have to restart it a few times).

**Review by zoq — COMMENTED** [review]
Might be a good idea to clean the commit history or we just do a squash & merge and clean it there.

File: .ci/windows-steps.yaml [hunk 20]
```diff
@@ -12,9 +12,14 @@ steps:
     nuget install boost -o $(Agent.ToolsDirectory) -Version 1.60.0
     nuget install boost_random-vc140 -o $(Agent.ToolsDirectory) -Version 1.60.0
     nuget install boost_math_c99-vc140 -o $(Agent.ToolsDirectory) -Version 1.60.0
-    nuget install OpenBLAS -o $(Agent.ToolsDirectory)
     nuget install unofficial-flayan-cereal -o $(Agent.ToolsDirectory)
-
+    nuget install ensmallen -o $(Agent.ToolsDirectory) -Version 2.17.0
+    ## Delete all ensmallen dependencies including armadillo headers, we do not need them here
```
Comment (zoq) [inline .ci/windows-steps.yaml:17]: I'm not sure I get the reason why we download the ensmallen package from nuget, and delete everything besides the header files. In the linux build we just download the header files and copy everything into the target directory. Is there anything that I'm missing here?

Comment (shrit) [reply .ci/windows-steps.yaml:17]: Frankly, I do not remember why we went into NuGet direction, there was a strong reason, but I can not remember why

Comment (shrit) [reply .ci/windows-steps.yaml:17]: After all of this suffering, I think we can merge this pull request. Whatever the first reason was, it is better to use NuGet for future versions, we will have to change only the number now for ensmallen

**Comment by rcurtin** [comment]
Still a memory error; I tried restarting the build again.

**Comment by shrit** [comment]
@rcurtin was it a linking error?

**Comment by rcurtin** [comment]
Nope, just 'compiler out of heap space'.  I'd just like to make sure we can at least get it to build cleanly once. :)

**Comment by shrit** [comment]
In this case, the issue will be coming from armadillo 9.8 since it is the first time we try to compile mlpack with it on Windows.
I do not know how we can avoid these, maybe cut the tests examples into several units?

**Comment by shrit** [comment]
All green :100: :rocket:

**Comment by rcurtin** [comment]
Awesome!  Great work @shrit, super cool we were finally able to get this one working right. :)

**Comment by shrit** [comment]
@rcurtin @jeffin143 thanks, I think we are close to merging #3006, if you have any spare time, do not hesitate in giving hints on the Unicode strange loop behaviors
